Month3/Project1/Project1/TDinamicArray.h

```diff
@@ -0,0 +1,290 @@
+#pragma once
+#include "Square.h"
+#include <stdlib.h>     
+#include <time.h> 
+
+
+template <typename T>
+class TDinamicArray
+{
+private:
+	int Capacity = 3;
+	int Size = 0;
+	T* Array = new T[Capacity];
+public:
+	TDinamicArray() 
+	{
+	
+	}
+	TDinamicArray(const int EntryCapacity)
+	{
+		Array = new T[EntryCapacity];
+	
+	}
+	~TDinamicArray() 
+	{
+		delete[] Array;
+	}
+
+	TDinamicArray(const TDinamicArray<T>& Array1)
+	{
+		Append(Array1);
+	}
+	
+
+	T& operator[](int Index)
+	{
+		return Array[Index];
+	}
+	const T& operator[](int Index)const
+	{
+		return Array[Index];
+	}
+
+
+
+	T& Front()
+	{
+		return Array[0];
+	}
+	const T& Front()const 
+	{
+		return Array[0];
+	}
+
+
+	T& Back()
+	{	
+		return Array[Size];
+	}
+	const T& Back() const
+	{
+		return Array[Size];
+	}
+
+
+	T& GetData()
+	{
+		return Array;
+	}
+	const T& GetData() const
+	{
+		return Array;
+	}
+
+	bool Empty()
+	{
+		return Capacity == 0;
+	}
+
+	int GetSize()
+	{
+		return Size;
+	}
+
+	int GetCapacity()
+	{
+		return Capacity;
+	}
+
+
+	void Reserve(int RequieredCapacity)
+	{
+		if (RequieredCapacity > Capacity)
+		{
+			T* NewArray = new T[RequieredCapacity];
+			for (int CurrentIndex = 0; CurrentIndex < Capacity; CurrentIndex++)
+			{
+				NewArray[CurrentIndex] = Array[CurrentIndex];
+			}
+			delete Array;
+
+			Capacity = RequieredCapacity;
+			Array = NewArray;
+		}
+
+		
+	}
+
+
+	void ShriktoFit()
+	{
+
+		T* NewArray = new T[Size];
+
+		for (int CurrentIndex = 0; CurrentIndex <= Size; CurrentIndex++)
+		{
+			NewArray[CurrentIndex] = Array[CurrentIndex];
+		}
+		delete Array;
+
+		Capacity = Size;
+		Array = NewArray;
+	}
+
+	void Clear()
+	{
+		if (Size > 0)
+		{
+			Size = 0;
+		}
+	}
+
+
+	void PushBack(const T& NewShape)
+	{
+		if (Capacity == 0)
+		{
+			Reserve(1);
+			Size++;
+			Array[Size] = NewShape;
+		}
+		if (Size == Capacity)
+		{
+			int NewCapacity = Capacity * 2;
+			T* NewArray = new T[NewCapacity];
+			for (int CurrentIndex = 0; CurrentIndex <= Capacity; CurrentIndex++)
+			{
+				NewArray[CurrentIndex] = Array[CurrentIndex];
+			}
+			Capacity = NewCapacity;
+			delete Array;
+			Array = NewArray;
+			Size += 1;
+			Array[Size] = NewShape;
+		}
+	}
+
+
+	void PopBack()
+	{
+		Size -= 1;
+	}
+
+	void Insert(int Index, const T& Shape)
+	{
+		std::cout << "In which position do you want to add your Shape?" << std::endl;
+		std::cin >> Index;
```

Comment:
Hm, once again, I'm not sure references to Shapes is appropriate for a generic Array. Furthermore, I think forcing an input from the user every time you want to insert is definitely not appropriate, as sometimes we as programmers would like to add an item to an array without input from the user.